src/providers/shapeshift/shapeshift.ts

```diff
@@ -162,13 +194,241 @@ export class ShapeshiftProvider {
+    let data = {
+      grant_type: 'refresh_token',
+      client_id: this.credentials.CLIENT_ID,
+      client_secret: this.credentials.CLIENT_SECRET,
+      redirect_uri: this.credentials.REDIRECT_URI,
+      refresh_token: refreshToken
+    };
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
+      Accept: 'application/json'
+    });
+
+    this.http.post(url, data, { headers }).subscribe(
+      data => {
+        this.logger.info('ShapeShift: Refresh Access Token SUCCESS');
+        this._afterTokenReceived(data, cb);
+      },
+      data => {
+        this.logger.error(
+          'ShapeShift: Refresh Access Token ERROR ' +
+            data.status +
+            '. ' +
+            this.getErrorsAsString(data.error)
+        );
+        return cb(data.error);
+      }
+    );
+  }
+
+  public getErrorsAsString(data): string {
+    let errData;
+
+    try {
+      if (data && data.errors) errData = data.errors;
+      else if (data && data.error) errData = data.error_description;
+      else return JSON.stringify(data);
+
+      if (!_.isArray(errData)) {
+        errData = errData && errData.message ? errData.message : errData;
+        return errData;
+      }
+
+      if (_.isArray(errData)) {
+        var errStr = '';
+        for (var i = 0; i < errData.length; i++) {
+          errStr = errStr + errData[i].message + '. ';
+        }
+        return errStr;
+      }
+
+      return JSON.stringify(errData);
+    } catch (e) {
+      this.logger.error(e);
+      return e;
+    }
+  }
+
+  public init = _.throttle(cb => {
+    if (_.isEmpty(this.credentials.CLIENT_ID)) {
+      return cb('ShapeShift is Disabled. Missing credentials.');
+    }
+    this.logger.debug('Trying to initialize ShapeShift...');
+
+    this.getStoredToken(accessToken => {
+      if (!accessToken) {
+        this.logger.debug('ShapeShift not linked');
+        return cb();
+      }
+      this.logger.debug('ShapeShift already has Token.');
+      this._getMainAccountId(accessToken, (err, accountId) => {
+        if (err) {
+          if (!err.errors) return cb(err);
+          if (err.errors && !_.isArray(err.errors)) return cb(err);
+
+          let expiredToken;
+          for (let i = 0; i < err.errors.length; i++) {
+            if (err.errors[i].id == 'expired_token') expiredToken = true;
+          }
+
+          if (expiredToken) {
+            this.logger.debug('Token Expired. Refresh and get new Token.');
+            this.persistenceProvider
+              .getShapeshiftRefreshToken(this.credentials.NETWORK)
+              .then(refreshToken => {
+                this._refreshToken(refreshToken, (err, newToken) => {
+                  if (err) return cb(err);
+                  this._getMainAccountId(newToken, (err, accountId) => {
+                    if (err) return cb(err);
+                    return cb(null, {
+                      accessToken: newToken,
+                      accountId
```

Comment:
Why we need accountId? `_getMainAccountId` should be used to verify that token is valid with `oauth/token/details`. `GetAccount` should be used separated for another purpose.